### ISSUE TITLE
"Pop out stage" (make stage appear in floating browser window/tab)

### ISSUE THREAD

**Issue by towerofnix** [ScratchAddons/ScratchAddons#1993]
**Describe the new addon**

Pulling from LLK/scratch-gui#2225: an addon which adds (probably near the fullscreen/presentation/small/normal-stage buttons) a button to make the stage into a separate browser tab. Plus, a button to move the stage back (or simply have the stage reappear when the popup window/tab is closed). This would allow for more particular split-screen views, or moving the stage to a different monitor/display.

I'm not sure how complicated this would be to implement, but I'm guessing it can be made a bit simpler by just making the stage area invisible (but still present) on the main page, and sending the canvas data over to the other tab. There would need to be a way to transfer events on that canvas (i.e. mouse/touch & keyboard input) over to the editor tab (with the VM etc), too!

**Comment by TheColaber** [comment]
Just a thought: If the addon will also be bringing the buttons like that flag and the stopsign, how will this affect addons that append there like the clone counter and the pause button?

Other than that, that would be cool...

**Comment by towerofnix** [comment]
@TheColaber That's a good point! It'd probably be more difficult than it's worth to somehow share those elements to the popup page. Maybe the addon would only hide the stage, but leave the actual stage controls in the editor window (maybe bringing just a mock/basic version of the green-flag so that control is easily accessible, and the "put back"/close pop-up button too, of course). Maybe a little counter-intuitive, but most of the controls are more needed while working in the editor anyway, so it could be handy!

**Comment by TheColaber** [comment]
We could have the popout be a certain URL that the addons will run on? I was thinking of the embed page but the project would need to have been saved then... Or we could just forget about the addons trying to run there...

**Comment by towerofnix** [comment]
Yeah, I thought about the embed page too, but then it disconnects the pop-out window from the editor altogether, which is no good. It's important that the pop-out window and the actual editor run on the same VM (and other components of the editor/player), so that it works as a way to put the stage elsewhere and give more room for the editor. That's the main motivation for this suggestion; for people who just want a more focused view on the player, there's already the embed and full-screen/presentation views :P

**Comment by towerofnix** [comment]
That's why I suggested the canvas implementation in OP, btw. It's the most compatible / least disruptive because it doesn't actually change the interface that Scratch (or addons) will be working with; it's just opening a new window with a canvas acting as another way to view and interact with the original canvas.

**Comment by Samq64** [comment]
>  it only supports <video> elements

🤦‍♂️ I should really think before commenting.

**Comment by Miala-python** [comment]
I don't know if it's possible, but can we make just a little window with scratch in full-screen mode who stays in front of the normal window ? (And possibly remove the flag bar…)

**Comment by WorldLanguages** [comment]
> What if this was implemented as a picture-in-picture window instead of a browser tab?

Chrome will soon provide a way to add arbitrary HTML content to a picture-in-picture window: https://developer.chrome.com/docs/web-platform/document-picture-in-picture/

**Comment by WorldLanguages** [comment]
I'll research this in a month or so.

**Comment by DNin01** [comment]
> I'll research this in a month or so.

Well, it's been about a month now.

For anyone with Microsoft PowerToys (v0.73.0+, I think?), this can be done manually with the "Crop and Lock" utility. Just thought I'd throw that out there.

**Comment by WorldLanguages** [comment]
I'm not sure if we can get this to work properly unless the user has a 2+ monitor set-up.

Unlike YouTube videos, projects need to run JavaScript each frame. And setInterval is very throttled when the tab is backgrounded. I've been doing some experiments, and Chrome doesn't really seem to care that the focused tab is a child of the main project tab, it is throttled anyway. Literally goes from 30FPS to 1FPS

**Comment by WorldLanguages** [comment]
I wonder what would happen if we add our own `setInterval()` call to the pop-out stage tab/window, and synchronously call the other tab's interval callback... Maybe that could work...

**Comment by WorldLanguages** [comment]
Looks like calling `setInterval()` from the pop-out-stage tab solves the issue in Chromium. Pretty simple!
Might need some compatibility work with the `60fps` addon which also messes with the vm runtime setInterval calls

**Comment by WorldLanguages** [comment]
Throttling issues (includes 60fps addon compatibility) were dealt with, see draft PR #6787

**Comment by Samq64** [comment]
Document picture-in-picture is now supported by both Chrome and Firefox.

**Comment by WorldLanguages** [comment]
Reminder: a draft PR is available at #6787. Feel free to fork it.
